src/library/itunes/itunesmacosimporter.h

```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include <QSqlDatabase>
+#include <QString>
+#include <atomic>
+
+#include "library/itunes/itunesimporter.h"
+#include "library/libraryfeature.h"
+
+class ITunesMacOSImporter : public ITunesImporter {
+  public:
+    ITunesMacOSImporter(LibraryFeature* parentFeature,
+            QSqlDatabase& database,
+            std::atomic<bool>& cancelImport);
```

Comment:
This is mostly an artifact from the existing iTunes XML importer which (even less safely) uses a non-atomic boolean for this. The `cancelImport` method + private member sounds good for now, I'd say.

Reply:
Hm, digging into this a bit it might require a bit of a rearchitect. Currently imports are cancelled by setting `m_cancelImport = true` in `~ITunesFeature`. Translating this into a call (e.g. to a new method `ITunesImporter::cancelImport`) would require us to store the `ITunesImporter` in a field instead during `ITunesFeature::importLibrary` which doesn't feel cleaner than just passing the reference to the importer.

I agree that passing a reference to private members is fragile and requires us to be very careful to have the importer not outlive its 'parent' feature. Wdyt about making this a `std::shared_ptr<std::atomic<bool>>` instead of a reference?

Reply:
> would require us to store the ITunesImporter in a field instead during ITunesFeature::importLibrary which doesn't feel cleaner than just passing the reference to the importer.

Well I agree that storing the somewhat temporary `ITunesImporter` is not great but neither is the current implicit contract around that `m_cancelImport`. I think temporarily storing the pointer to the importer object is still an improvement.

> Then perhaps keeping it the current way and refactoring it in a future PR might be the best way to go about this, considering that it would also be deviation from the way this is currently handled in the Banshee/Traktor/Rhythmbox/etc. integration?


Yes, that definitely an option. Could you then revert those changes and instead leave a TODO summarizing what we've discussed so far? So primarily that we should add a **thread-safe** cancel function to the `ITunesImporter` interface (possibly to `BaseExternalLibraryFeature` as well) and that `m_cancelImport` needs to be made atomic everywhere.

Reply:
The deviation I meant was w.r.t migrating the 'plain' boolean flag approach to a cancel method.

> Could you then revert those changes

Which ones specifically? 7d9ac9bd1d36d35d9b80f00ebc44e464a79aee8d? If we are going to use atomics in the future anyway, I don't see much harm in keeping it (from my understanding `bool&` is thread-unsafe, while passing `std::atomic<bool>&` is safe, but fragile/not very pretty due to passing references to private members around).

Reply:
I've added some notes in 48250c10c4e2e065b4fd56b0ea6bddf6c50c161c as you suggested